syft/pkg/cataloger/java/parse_pom_xml.go

```diff
@@ -235,59 +235,66 @@ func cleanDescription(original *string) (cleaned string) {
 func resolveProperty(pom gopom.Project, property *string, propertyName string) string {
 	propertyCase := safeString(property)
 	log.WithFields("existingPropertyValue", propertyCase, "propertyName", propertyName).Trace("resolving property")
-	return propertyMatcher.ReplaceAllStringFunc(propertyCase, func(match string) string {
-		propertyName := strings.TrimSpace(match[2 : len(match)-1]) // remove leading ${ and trailing }
-		entries := pomProperties(pom)
-		if value, ok := entries[propertyName]; ok {
-			return value
-		}
+	for propertyMatcher.MatchString(propertyCase) {
+		oldPropertyCase := propertyCase
+		propertyCase = propertyMatcher.ReplaceAllStringFunc(propertyCase, func(match string) string {
+			propertyName := strings.TrimSpace(match[2 : len(match)-1]) // remove leading ${ and trailing }
+			entries := pomProperties(pom)
+			if value, ok := entries[propertyName]; ok {
+				return value
```

Comment:
Instead of having a loop, adding a recursive call here to `resolveProperty` could also be a solution?

Reply:
That diff would potentially be a _lot_ easier to read. Let me give that a shot. Thanks for the suggestion!

Reply:
Actually, in adding new tests for this implementation, I realized that there are two cases we need to make sure halt:

1. Non-rooted deference, where following the path of repeated de-reference ends in a variable
2. Cyclic dereference, where a variable refers to itself

Neither of these are well-formed pom.xml, we just need to defend against them hanging or overflowing the stack or something.